File: 백준/Gold/1806. 부분합/부분합.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
int n, m;

int arr[100000];
int main() {

	ios::sync_with_stdio(0);
	cin.tie(0);
	cin >> n >> m;
	for (int i = 0; i < n; i++) {
		cin >> arr[i];
	}
	int st, en;
	st = 0;
	en = 0;
	int tot = arr[st];
	
	int ans = 987654321;
	//길이는 en-st+1
	while (st < n && en < n) {
		if (st == en && arr[st] >= m) {
			ans = 1;
			break;
		}
		if (tot < m) {
			en++;
			tot += arr[en];
		}
		else {
			ans = min(ans, en - st + 1);
			tot -= arr[st];
			st++;
		}
	}
	
	if (ans == 987654321) ans = 0;
	cout << ans;

}

File: 백준/Gold/1806. 부분합/부분합.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	
	static int N, S, len, ans;
	static int arr[];
	static long sum;
	public static void main(String[] args) throws IOException {
		BufferedReader br =new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N=Integer.parseInt(st.nextToken());
		S=Integer.parseInt(st.nextToken());
		arr=new int[N+1];
		st=new StringTokenizer(br.readLine());
		for(int i=0;i<N;i++) {
			arr[i]=Integer.parseInt(st.nextToken());
		}
		ans=Integer.MAX_VALUE; //가장 긴 길이 
		int s=0;
		int e=0; 
		sum=arr[0]; //시작 
		while(e<N && s<=e) {
			if(sum<S) {
				//아직 누적합보다 작다 
				e++;
				sum+=arr[e]; 
			}
			else {
				//누적합보다 크다 
				len=e-s+1;
				ans=Math.min(ans, len); //길이가 더 작다면 갱신 
				sum-=arr[s];
				s++;
			}
		}
		if(ans==Integer.MAX_VALUE) ans=0;
		System.out.println(ans);
	}

}
